### DISCUSSION TITLE
rclcpp::Node usage and patterns

### DISCUSSION THREAD

**Discussion by tylerjw** [moveit/moveit2#1489] [General]
Going through the parameter interfaces in MoveIt lead me to ask the question, what are the "nodes" in MoveIt that users create and interact with, and what are the "Nodes" we create and interact with?

MoveIt has many usages of rclcpp::Node and each has various reasons for existing. I will look into first documenting the current usage and considering if we can reduce the complexity and standardize our usage of `rclcpp::Node`.

**Comment by gavanderhoorn** [comment]
Did you intend to open this discussion on the `ros-planning` organisation instead of the MoveIt2 discussion board?

**Reply by tylerjw** [reply]
I think we set this up in the `ros-planning` org to invite more people to participate. I am generally working on MoveIt but have been reading nav2's implementation to try to learn from the patterns of usage in nav2.

**Reply by tylerjw** [reply]
Right now imo MoveIt is a real mess when it comes to how its public and not-so-public ros/c++ interfaces are organized along with its usage of ROS interfaces. I would like to find ways to simplify and standardize what MoveIt does so it can be easier to use and have more well-defined interfaces and patterns for use.

**Comment by schornakj** [comment]
I've noticed lots of places (in MoveIt 2 primarily) where we create internal instances of `rclcpp::Node`. I think this is an antipattern because it blurs the boundary between "rclcpp Executor land" and "business logic callback land" and opens up more opportunities for callback group issues and thread lifetime bugs.
